spec.html

```diff
@@ -42415,104 +42525,219 @@ <h1>
       </emu-clause>
 
       <emu-clause id="sec-notifywaiter" type="abstract operation">
         <h1>
           NotifyWaiter (
-            _WL_: a WaiterList,
-            _W_: an agent signifier,
+            _WL_: a WaiterList Record,
+            _waiterRecord_: a Waiter Record,
           ): ~unused~
         </h1>
         <dl class="header">
         </dl>
         <emu-alg>
           1. Assert: The surrounding agent is in the critical section for _WL_.
-          1. Notify the agent _W_.
+          1. Assert: _waiterRecord_.[[Result]] is either *"ok"* or *"timed-out"*.
+          1. If _waiterRecord_.[[PromiseCapability]] is ~blocking~, then
+            1. Wake the agent whose signifier is _waiterRecord_.[[AgentSignifier]] from suspension.
+            1. NOTE: This causes the agent to resume execution in SuspendThisAgent.
+          1. Else if AgentSignifier() is _waiterRecord_.[[AgentSignifier]], then
+            1. Let _promiseCapability_ be _waiterRecord_.[[PromiseCapability]].
+            1. Perform ! Call(_promiseCapability_.[[Resolve]], *undefined*, « _waiterRecord_.[[Result]] »).
+          1. Else,
+            1. Perform HostEnqueueResolveInAgentJob(_waiterRecord_.[[AgentSignifier]], _waiterRecord_.[[PromiseCapability]], _waiterRecord_.[[Result]]).
+            1. NOTE: An agent must not access another agent's promise capability in any capacity beyond passing it to the host.
           1. Return ~unused~.
         </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-hostenqueueresolveinagentjob" type="host-defined abstract operation">
+        <h1>
+          HostEnqueueResolveInAgentJob (
+            _agentSignifier_: an agent signifier,
+            _promiseCapability_: a PromiseCapability Record,
+            _resolution_: an ECMAScript language value,
+          ): ~unused~
+        </h1>
+        <dl class="header">
+        </dl>
+        <p>Let _resolveJob_ be a new Job Abstract Closure with no parameters that captures _agentSignifier, _promiseCapability_, and _resolution_ and performs the following steps when called:</p>
+        <emu-alg>
+          1. Assert: AgentSignifier() is _agentSignifier_.
+          1. Perform ! Call(_promiseCapability_.[[Resolve]], *undefined*, « _resolution_ »).
+          1. Return ~unused~.
+        </emu-alg>
+        <p>An implementation of HostEnqueueResolveInAgentJob must schedule _resolveJob_ in the agent signified by _agentSignifier_. The host may delay performing _resolveJob_ in the agent signified by _agentSignifier_, e.g. for resource management reasons, but it eventually must be performed for forward progress.</p>
+        <p>An implementation of HostEnqueueResolveInAgentJob must conform to the requirements in <emu-xref href="#sec-jobs"></emu-xref>.</p>
+      </emu-clause>
+
+      <emu-clause id="sec-dowait" type="abstract operation">
+        <h1>
+          DoWait (
+            _mode_: ~sync~ or ~async~,
+            _typedArray_: a TypedArray,
```

Comment:
```suggestion
            _typedArray_: an ECMAScript language value
```
because the caller isn't ensuring it's a TypedArray.